src/IO/ReadBufferFromS3.cpp

```diff
@@ -45,20 +45,27 @@ bool ReadBufferFromS3::nextImpl()
 {
     Stopwatch watch;
     bool next_result = false;
-    auto sleep_time_with_backoff_milliseconds = std::chrono::milliseconds(100);
 
-    if (!impl)
+    if (impl)
+    {
+        /// `impl` has been initialized earlier and now we're at the end of the current portion of data.
+        impl->position() = position();
```

Comment:
Without this change the test [test_encrypted_disk/test.py::test_encrypted_disk[s3_policy]](https://github.com/ClickHouse/ClickHouse/pull/24206/files#diff-810f8625742ff9db0e7c7950bcbdbca1230f86ecedf7156fa18b4e017591776cR23) failed with assertion

```
../src/IO/ReadBuffer.h:58: bool DB::ReadBuffer::next(): Assertion `!hasPendingData()' failed
```